Get HTML file of with code coverage info

ArtiomDivak opened the pull request:
By using lcov and genhtml inside each container we get res directory.
Inside the res directory will be index.html that will contain all the
info about the code coverage

mwperina commented on tests/bluechi_test/container.py:
```diff
@@ -67,8 +67,34 @@ def gather_coverage(self, data_coverage_dir: str) -> None:
+
+        result_file_name = f"res-{self.container.name}"
+
+        self.container.exec_run(f"genhtml coverage.info --output-directory={result_file_name}")
```
The above code just generates HTML output for a single test, which is not that useful in a global point of view. Is there a way how to merge those `.info` files and generate global HTML output from all tests?

mwperina replied:
Maybe something like https://stackoverflow.com/a/60636596

engelmi replied:
> The above code just generates HTML output for a single test, which is not that useful in a global point of view.

Wouldn't we want to have both cases? Reports for each test might be useful for error tracking and one combining the results from all as overview.

engelmi reviewed (commented):
I am not sure generating the report inside the test and container is good. I think just collecting the coverage data files and exporting them is enough logic inside the test - the report can then be generated in the github `test` job after the step running the tests.

ArtiomDivak committed:
Get Info files of tests

Get Info file file from each container of each test.
This info files will be merged with lcov command.

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/397
Signed-off-by: Artiom Divak <adivak@redhat.com>

mwperina commented on tests/bluechi_test/container.py:
```diff
@@ -67,8 +67,25 @@ def gather_coverage(self, data_coverage_dir: str) -> None:
+        bluechi_version = "bluechi" + str(output.decode()).split("-controller")[1].split("\n")[0]
+        src_file_location = f"/usr/src/debug/{bluechi_version}/src"
+
+        self.container.exec_run(f"cp -r {gcno_file_location}/. {gcda_file_location}")
```
Just na idea: Wouldn't it be easier to package source files inside `bluechi-coverage` package instead of this complicated copy of source files into `.gcno` files location?

mwperina commented on tests/tests/tier0/z-bluechi-coverage-to-html/test_z_bluechi_coverage_to_html.py:
```diff
@@ -0,0 +1,51 @@
+
+def exec(ctrl: BluechiControllerContainer, nodes: Dict[str, BluechiNodeContainer]):
+    path_to_tests_results = os.popen(
+        "echo $TMT_TREE").read().split('/tree')[0] + "/execute/data/guest/default-0/tests/tier0/"
```
Please use `os.environ.get` to fetch the value of `TMT_TREE`

ArtiomDivak committed:
Added finish step script to make html from info files

Running script when WITH_COVERAGE is 1. This script will run after all
tests and will get an html report for the coverage from all the info
files.

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/397
Signed-off-by: Artiom Divak <adivak@redhat.com>

mwperina commented on tests/plans/tier0.fmf:
```diff
@@ -21,4 +21,9 @@ prepare:
+    - name: Create coverage report
+      how: shell
+      script: |
+          bash scripts/create_coverag_report.sh
```
Can't we just run following command directly?

```
- name: Create coverage report
  how: shell
  script: |
      pytest -svv scripts/create_coverag_report.py
```

mwperina commented on tests/scripts/create_coverag_report.py:
```diff
@@ -0,0 +1,51 @@
+
+def exec(ctrl: BluechiControllerContainer, nodes: Dict[str, BluechiNodeContainer]):
+    path_to_tests_results = os.popen(
+        "echo $TMT_TREE").read().split('/tree')[0] + "/execute/data/guest/default-0/tests/tier0/"
```
Can't we use `os.environ` python function to get `TMT_TREE` variable value into python code?

mwperina commented on tests/scripts/create_coverag_report.sh:
We run the create_coverage_report.py directly, could you please remove this shell script?

mwperina commented on tests/scripts/create_coverag_report.py:
Please rename to  `create_coverage_report.py`

mwperina commented on tests/containers/integration-test-local:
```diff
@@ -4,6 +4,12 @@ RUN mkdir -p /tmp/bluechi-rpms
+
+RUN dnf install --nodocs \
+        lcov \
+    -y 
```
Could you please move this into integration-test-base container?

mwperina commented on tests/README.md:
````diff
@@ -86,6 +86,16 @@ tmt run -v -eWITH_VALGRIND=1
+tmt run -v -eWITH_COVERAGE=1
+```
+
+After the integration tests will finish the code coverage html result can be found in 'res' dir inside the tmt execution result directory.
````
Could you please add concrete path as an example?

mwperina commented on tests/README.md:
```diff
@@ -86,6 +86,16 @@ tmt run -v -eWITH_VALGRIND=1
 
 If `valgrind` detects a memory leak in a test, the test will fail, and the logs will be found in the test `data` directory.
 
+## Running integration tests with code coverage
```
I'd change to *Running integration tests with code coverage report*

mwperina commented on tests/README.md:
```diff
@@ -86,6 +86,16 @@ tmt run -v -eWITH_VALGRIND=1
 
+## Running integration tests with code coverage
+
+To run integration tests with `code coverage`, set `WITH_COVERAGE` environment variable as follows:
```
I'd change the sentence to *To produce code coverage report during integration tests run set `WITH_COVERAGE` environment variable as follows:*

mwperina reviewed (commented):
Some small comments around doc, but other than that LGTM

mwperina commented on tests/README.md:
```diff
@@ -86,6 +86,18 @@ tmt run -v -eWITH_VALGRIND=1
 
+## Running integration tests with code coverage report
+
+To produce code coverage report during integration tests run set WITH_COVERAGE environment variable as follows:
```
Could you please either  mention that RPMs need to be built with coverage support to be able to run integration test with coverage report or move section about build RPMs with coverage support into this section

mwperina commented on tests/scripts/create_coverage_report.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def exec(ctrl: BluechiControllerContainer, nodes: Dict[str, BluechiNodeContainer]):
+    path_to_info_files = os.environ["TMT_TREE"].split('/tree')[0] + "/execute/data/guest/default-0/tests/tier0/"
```
Can't we limit the path to "/execute/data" to minimize the risk that the path could change if tests are executed for example on multiple hosts?

mwperina reviewed (approved):
LGTM

mwperina committed:
Adds logging to gathering code coverage artifacts

Related-to: https://github.com/eclipse-bluechi/bluechi/issues/397
Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi reviewed (approved):
LGTM
Could you fetch the latest changes before merging? @ArtiomDivak
